### PR TITLE
backport: add Events method to the client interface (#7982)

### PR DESCRIPTION
**Pull request by mmsqe**
For the full description, see https://github.com/tendermint/tendermint/pull/7982

---

#### PR checklist

- [ ] Tests written/updated, or no tests needed
- [ ] `CHANGELOG_PENDING.md` updated, or no changelog entry needed
- [ ] Updated relevant documentation (`docs/`) and code comments, or no
      documentation updates needed

### CODE REVIEW

File: rpc/client/interface.go [hunk 1]
```diff
@@ -115,20 +116,40 @@ type NetworkClient interface {
...
-	// cap=1 onto which events are published. An error is returned if it fails to
-	// subscribe. outCapacity can be used optionally to set capacity for the
-	// channel. Channel is never closed to prevent accidental reads.
+	// Events fetches a batch of events from the server matching the given query
+	// and time range.
+	Events(ctx context.Context, req *ctypes.RequestEvents) (*ctypes.ResultEvents, error)
+}
+
+// TODO(creachadair): This interface should be removed once the streaming event
+// interface is removed in Tendermint v0.37.
```
Comment (cmwaters) [inline rpc/client/interface.go:127]:
```suggestion
// interface is removed in Tendermint v0.39.
```
nit

File: light/proxy/routes.go [hunk 2]
```diff
@@ -300,3 +304,31 @@ func makeBroadcastEvidenceFunc(c *lrpc.Client) rpcBroadcastEvidenceFunc {
...
+
+type rpcEventsSearchFunc func(
+	ctx *rpctypes.Context,
+	filter string,
+	maxItems int,
+	before, after string,
+	waitTime time.Duration,
+) (*ctypes.ResultEvents, error)
+
+func makeEventsSearchFunc(c *lrpc.Client) rpcEventsSearchFunc {
```
[the remark this answers is not in the document]
Comment (cmwaters) [reply light/proxy/routes.go:316]: Do you know which PR the json parser refactor was?

**Review by cmwaters — APPROVED** [review]

File: light/proxy/routes.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (mmsqe) [reply light/proxy/routes.go:316]: Mainly miss [parseParams](https://github.com/tendermint/tendermint/pull/7670/files?diff=split&w=0#diff-887d277f79c80252e9460a9d1fee2e0543b942ca0dab36258f3580cb2c95d34dL147).

[the remark this answers is not in the document]
Comment (cmwaters) [reply light/proxy/routes.go:316]: ok, thanks. Noted it as an issue

File: rpc/client/local/local.go [hunk 3]
```diff
@@ -127,6 +127,10 @@ func (c *Local) ConsensusParams(ctx context.Context, height *int64) (*ctypes.Res
 	return core.ConsensusParams(c.ctx, height)
 }
 
+func (c *Local) Events(ctx context.Context, req *ctypes.RequestEvents) (*ctypes.ResultEvents, error) {
+	return core.Events(c.ctx, req.Filter.Query, req.MaxItems, req.Before, req.After, req.WaitTime)
```
Comment (sergio-mena) [inline rpc/client/local/local.go:131]: Out of curiosity: #7982 implements the new functionality directly in this file, whereas this PR moves it inside the `core` package. Was there a reason?

**Review by sergio-mena — APPROVED** [review]
LGTM

File: rpc/client/local/local.go (hunk 3 above)
Comment (mmsqe) [reply rpc/client/local/local.go:131]: To allow reuse in here and [routes](https://github.com/tendermint/tendermint/blob/0dce59e60d936e8e35b09d7ab43cc7032a417c09/rpc/core/routes.go#L13), while https://github.com/tendermint/tendermint/pull/7982 with parseParams refactor only use in [local](https://github.com/tendermint/tendermint/blob/5deb9689ff6deebf43e13cae618785c18d751c1c/rpc/client/local/local.go#L133).

**Pull request merged**